src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs

```diff
@@ -567,6 +548,48 @@ impl<'tcx> GotocCtx<'tcx> {
         )
     }
 
+    /// Generates either a panic or no-op for `assert_*` intrinsics.
+    /// These are intrinsics that statically compile to panics if the type
+    /// layout is invalid so we get a message that mentions the offending type.
+    fn codegen_assert_intrinsic(
+        &mut self,
+        instance: Instance<'tcx>,
+        intrinsic: &str,
+        span: Option<Span>,
+    ) -> Stmt {
+        let ty = instance.substs.type_at(0);
+        let layout = self.layout_of(ty);
+
+        // For all intrinsics we first check `is_uninhabited` to give a more
+        // precise error message
+        if layout.abi.is_uninhabited() {
+            return self.codegen_fatal_error(
+                &format!("attempted to instantiate uninhabited type `{}`", ty),
+                span,
+            );
+        }
+
+        // Then we check if the type allows "raw" initialization for the cases
+        // where memory is zero-initialized or entirely uninitialized
+        if intrinsic == "assert_zero_valid" && !layout.might_permit_raw_init(self, true) {
+            return self.codegen_fatal_error(
+                &format!("attempted to zero-initialize type `{}`, which is invalid", ty),
+                span,
+            );
+        }
+
+        if intrinsic == "assert_uninit_valid" && !layout.might_permit_raw_init(self, false) {
```

Comment:
Is there a spec we can point to here?

Reply:
I was not able to find one as documentation regarding these intrinsics is quite hard to find.

The closest thing I could find is the actual implementation in `rustc_codegen_ssa/src/mir/block.rs` (`codegen_panic_intrinsic`) in addition to PR discussions. But if you know of one I will be happy to add it here.

Reply:
FWIW, I think links directly to code serve as a reasonable reference, too. Maybe add that as a comment? (e.g. link directly to it on github?)

Reply:
My main concern with that is that links to `rustc` code may become outdated pretty quickly. But even that is better that no links, so just added a note here linking to that function.